File: api/src/pipeline/stages/research.py
"""Research stage: Perplexity API for keyword research."""

from __future__ import annotations

import logging
import re

from src.pipeline.helpers import (
    StageTimer,
    build_stage_prompt,
    load_rules,
    publish_stage_log,
)
from src.pipeline.state import PipelineState
from src.services.llm import LLMResponse, PerplexityClient

logger = logging.getLogger(__name__)

# Phrases that indicate Perplexity returned a meta-response instead of research
_REFUSAL_PATTERNS = [
    r"I'm\s+\*?\*?Perplexity\*?\*?",
    r"I(?:'m| am) a search assistant",
    r"I need to clarify my role",
    r"I(?:'m| am) not a blog research agent",
    r"I can(?:not|'t) (?:create|execute|generate|access)",
    r"What I \*?can\*? do instead",
    r"To move forward",
    r"Please provide (?:either|new search results)",
    r"Which would be most helpful\?",
]
_REFUSAL_RE = re.compile("|".join(_REFUSAL_PATTERNS), re.IGNORECASE)

# Sections we expect in valid research output
_EXPECTED_SECTIONS = ["keyword", "pain point", "competitor", "search intent"]

MAX_RESEARCH_ATTEMPTS = 3


def _is_valid_research(content: str) -> bool:
    """Return True if the response looks like actual research, not a meta-response."""
    if _REFUSAL_RE.search(content):
        return False
    content_lower = content.lower()
    matches = sum(1 for s in _EXPECTED_SECTIONS if s in content_lower)
    if matches < 2:
        return False
    return True


async def research_node(state: PipelineState) -> dict:
    """Execute the research stage using Perplexity API."""
    logger.info(f"Research stage starting for post {state.get('post_id')}")

    rules = load_rules("research")
    await publish_stage_log("Rules loaded, building prompt...", stage="research")
    prompt = build_stage_prompt("research", rules, state)

    system_msg = (
        "You are an expert SEO content researcher. "
        "Respond ONLY with the research document in markdown format. "
        "Do not discuss your capabilities or ask clarifying questions. "
        "Produce the complete research directly."
    )

    client = PerplexityClient(api_key=state.get("api_keys", {}).get("perplexity"))
    try:
        response: LLMResponse | None = None
        total_tokens_in = 0
        total_tokens_out = 0
        total_duration = 0.0

        for attempt in range(1, MAX_RESEARCH_ATTEMPTS + 1):
            msg = (
                f"Calling Perplexity sonar-pro "
                f"(attempt {attempt}/{MAX_RESEARCH_ATTEMPTS})..."
                if attempt > 1
                else "Calling Perplexity sonar-pro..."
            )
            await publish_stage_log(msg, stage="research")

            with StageTimer() as timer:
                response = await client.chat(
                    prompt=prompt if attempt == 1 else _reinforced_prompt(prompt),
                    system=system_msg,
                )

            total_tokens_in += response.tokens_in
            total_tokens_out += response.tokens_out
            total_duration += timer.duration

            if _is_valid_research(response.content):
                break

            logger.warning(
                f"Research attempt {attempt} returned meta-response, "
                f"retrying... (tokens: {response.tokens_out})"
            )
            await publish_stage_log(
                f"Response validation failed (attempt {attempt}), retrying...",
                stage="research",
                level="warning",
            )
        else:
            # All attempts returned invalid responses — use last response but log error
            logger.error(
                "All research attempts returned meta-responses. "
                "Using last response as fallback."
            )
            await publish_stage_log(
                "WARNING: Research quality may be degraded — "
                "Perplexity returned unexpected responses.",
                stage="research",
                level="error",
            )
    finally:
        await client.close()

    assert response is not None  # guaranteed by loop

    await publish_stage_log(
        f"Received {total_tokens_out} tokens in {total_duration:.1f}s",
        stage="research",
    )

    meta = {
        "stage": "research",
        "model": response.model,
        "tokens_in": total_tokens_in,
        "tokens_out": total_tokens_out,
        "duration_s": total_duration,
    }

    return {
        "research": response.content,
        "current_stage": "research",
        "stage_status": {
            **state.get("stage_status", {}),
            "research": "complete",
        },
        "_stage_meta": meta,
    }


def _reinforced_prompt(original_prompt: str) -> str:
    """Wrap the original prompt with stronger instructions for retry attempts."""
    return (
        "IMPORTANT: You must respond with ONLY the research document content. "
        "Do NOT describe yourself, your limitations, or ask questions. "
        "Do NOT say you are Perplexity or a search assistant. "
        "Simply produce the research document as specified.\n\n" + original_prompt
    )

File: web/src/mastra/steps/research.ts
/**
 * The `research` stage as a Mastra step, ported from `research_node` in
 * `api/src/pipeline/stages/research.py`.
 *
 * The step owns everything around the provider call: reading the run's state
 * from the posts table, assembling the prompt from `rules/blog-research.md`,
 * the meta-response retry loop, and committing the output to
 * `posts.research_content` before it returns. The agent (`agents/research.ts`)
 * owns only the model, the system message and the credential.
 *
 * The retry loop is the reason this stage is not a bare `createStep(agent)`.
 * Perplexity is a search assistant by default and answers a research brief with
 * a description of itself often enough that Python grew a validator and two
 * reinforced retries around it; dropping that would quietly regress research
 * quality on exactly the runs that already fail.
 */
import { createStep } from "@mastra/core/workflows/evented"

import { loadPipelineState, saveStageOutput } from "../post-state"
import { buildStagePrompt, loadRules } from "../prompts"
import { STATUS_COMPLETE } from "../state"
import {
  skippedStageOutput,
  shouldRunStage,
  stageStepInputSchema,
  stageStepOutputSchema,
} from "./stage-io"

/** Python's `MAX_RESEARCH_ATTEMPTS`. */
export const MAX_RESEARCH_ATTEMPTS = 3

/**
 * Phrases that mark a meta-response rather than research, from Python's
 * `_REFUSAL_PATTERNS`. Kept as the same alternation under the same
 * case-insensitive flag so both stacks reject the same responses; the patterns
 * use no Python-only regex syntax, so they port across verbatim.
 */
export const REFUSAL_PATTERNS = [
  String.raw`I'm\s+\*?\*?Perplexity\*?\*?`,
  String.raw`I(?:'m| am) a search assistant`,
  String.raw`I need to clarify my role`,
  String.raw`I(?:'m| am) not a blog research agent`,
  String.raw`I can(?:not|'t) (?:create|execute|generate|access)`,
  String.raw`What I \*?can\*? do instead`,
  String.raw`To move forward`,
  String.raw`Please provide (?:either|new search results)`,
  String.raw`Which would be most helpful\?`,
]

const REFUSAL_RE = new RegExp(REFUSAL_PATTERNS.join("|"), "i")

/** Sections a valid research document is expected to cover. */
export const EXPECTED_SECTIONS = ["keyword", "pain point", "competitor", "search intent"]

/**
 * True when the response looks like research rather than a meta-response:
 * no refusal phrasing, and at least two of the expected sections mentioned.
 */
export function isValidResearch(content: string): boolean {
  if (REFUSAL_RE.test(content)) return false
  const lower = content.toLowerCase()
  const matches = EXPECTED_SECTIONS.filter((section) => lower.includes(section)).length
  return matches >= 2
}

/** Python's `_reinforced_prompt`: the retry preamble, byte-identical. */
export function reinforcedPrompt(prompt: string): string {
  return (
    "IMPORTANT: You must respond with ONLY the research document content. " +
    "Do NOT describe yourself, your limitations, or ask questions. " +
    "Do NOT say you are Perplexity or a search assistant. " +
    "Simply produce the research document as specified.\n\n" + prompt
  )
}

export const researchStep = createStep({
  id: "research",
  inputSchema: stageStepInputSchema,
  outputSchema: stageStepOutputSchema,
  execute: async ({ inputData, mastra }) => {
    const { postId } = inputData
    const state = await loadPipelineState(postId)
    if (!shouldRunStage("research", inputData, state.stageStatus)) {
      return skippedStageOutput(inputData, "research")
    }
    const prompt = buildStagePrompt("research", loadRules("research"), state)
    const agent = mastra.getAgent("research")

    let text = ""
    let model = ""
    let tokensIn = 0
    let tokensOut = 0
    let durationMs = 0

    for (let attempt = 1; attempt <= MAX_RESEARCH_ATTEMPTS; attempt += 1) {
      const startedAt = Date.now()
      const result = await agent.generate(attempt === 1 ? prompt : reinforcedPrompt(prompt))
      durationMs += Date.now() - startedAt

      text = result.text
      model = result.response?.modelId ?? model
      // Python sums usage across every attempt, so a run that retried is billed
      // and reported for all of its calls rather than only the surviving one.
      tokensIn += result.usage?.inputTokens ?? 0
      tokensOut += result.usage?.outputTokens ?? 0

      if (isValidResearch(text)) break

      mastra.getLogger()?.warn("research attempt returned a meta-response, retrying", {
        postId,
        attempt,
        maxAttempts: MAX_RESEARCH_ATTEMPTS,
      })
    }

    if (!isValidResearch(text)) {
      // Python keeps the last response rather than failing the run: degraded
      // research still lets a human read the post detail page and rerun the
      // stage, where a hard failure would leave the column empty.
      mastra.getLogger()?.error("all research attempts returned meta-responses, using the last", {
        postId,
      })
    }

    await saveStageOutput(postId, "research", text, {
      ...state.stageStatus,
      research: STATUS_COMPLETE,
    })

    return {
      postId,
      stages: inputData.stages,
      stage: "research" as const,
      skipped: false,
      model,
      tokensIn,
      tokensOut,
      durationS: durationMs / 1000,
    }
  },
})
